src/mpfao/3D/tdympfao3D_core.c

```diff
@@ -979,3 +1032,317 @@ PetscErrorCode TDyComputeTransmissibilityMatrix3DMesh(TDy tdy) {
+  if (cell_id_up < 0 && cell_id_dn < 0) {
+    SETERRQ(PETSC_COMM_WORLD, PETSC_ERR_USER, "Both cell IDs sharing a face are not valid");
+  }
+
+  PetscInt dim = 3;
+  PetscReal coord_up[dim], coord_dn[dim], coord_face[dim];
+
+  ierr = TDyFace_GetCentroid(faces, face_id, dim, &coord_face[0]);
+  CHKERRQ(ierr);
+
+  if (cell_id_up >= 0) {
+    ierr = TDyCell_GetCentroid2(cells, cell_id_up, dim, &coord_up[0]); CHKERRQ(ierr);
+  } else {
+    for (PetscInt d = 0; d < 3; d++) coord_up[d] = coord_face[d];
+  }
+
+  if (cell_id_dn >= 0) {
+    ierr = TDyCell_GetCentroid2(cells, cell_id_dn, dim, &coord_dn[0]); CHKERRQ(ierr);
+  } else {
+    for (PetscInt d = 0; d < 3; d++) coord_dn[d] = coord_face[d];
+  }
+
+  ierr = TDyUnitNormalVectorJoiningTwoVertices(coord_up, coord_dn, up2dn_uvec); CHKERRQ(ierr);
+
+  PetscFunctionReturn(0);
+}
+
+/* -------------------------------------------------------------------------- */
+/// Computes upwind and downwind distance of cells sharing a face. If the face is a
+/// boundary face, one of the distance is zero
+///
+/// @param [in] tdy A TDy struct
+/// @param [in] face_id ID of the face
+/// @param [out] *dist_up Distance between the upwind cell centroid and face centroid
+/// @param [out] *dist_dn Distance between the downwind cell centroid and face centroid
+/// @returns 0 on success, or a non-zero error code on failure
+PetscErrorCode ComputeUpAndDownDist(TDy tdy, PetscInt face_id, PetscReal *dist_up, PetscReal *dist_dn) {
+
+  PetscFunctionBegin;
+
+  TDyMesh *mesh = tdy->mesh;
+  TDyCell *cells = &mesh->cells;
+  TDyFace *faces = &mesh->faces;
+  PetscErrorCode ierr;
+
+  PetscInt fOffsetCell = faces->cell_offset[face_id];
+  PetscInt cell_id_up = faces->cell_ids[fOffsetCell + 0];
+  PetscInt cell_id_dn = faces->cell_ids[fOffsetCell + 1];
+
+  if (cell_id_up < 0 && cell_id_dn < 0) {
+    SETERRQ(PETSC_COMM_WORLD, PETSC_ERR_USER, "Both cell IDs sharing a face are not valid");
+  }
+
+  PetscInt dim = 3;
+
+  PetscReal coord_face[dim];
+  ierr = TDyFace_GetCentroid(faces, face_id, dim, &coord_face[0]); CHKERRQ(ierr);
+
+  if (cell_id_up >= 0) {
+    PetscReal coord_up[dim];
+    ierr = TDyCell_GetCentroid2(cells, cell_id_up, dim, &coord_up[0]); CHKERRQ(ierr);
+    ierr = TDyComputeLength(coord_up, coord_face, dim, dist_up); CHKERRQ(ierr);
+  } else {
+    *dist_up = 0.0;
+  }
+
+  if (cell_id_dn >= 0) {
+    PetscReal coord_dn[dim];
+    ierr = TDyCell_GetCentroid2(cells, cell_id_dn, dim, &coord_dn[0]); CHKERRQ(ierr);
+    ierr = TDyComputeLength(coord_dn, coord_face, dim, dist_dn); CHKERRQ(ierr);
+  } else {
+    *dist_dn = 0.0;
+  }
+
+  PetscFunctionReturn(0);
+}
+
+/* -------------------------------------------------------------------------- */
+/// Computes face permeability tensor as a harmonically distance-weighted
+//  permeability of upwind and downwind permeability tensors
+///
+/// @param [in] tdy A TDy struct
+/// @param [in] face_id ID of the face
+/// @param [out] Kup[9] upwind permeability tensor
```

Comment:
Let's remove [9] and describe this and the following line as "components of the upwind/downwind permeability tensor"

Reply:
and maybe indicate whether the components are stored in row-major or column-major order.